File: app/src/main/java/com/dpis/module/home/HomeActivationStateResolver.java
package com.dpis.module.home;

import io.github.libxposed.service.XposedService;

public final class HomeActivationStateResolver {
    private HomeActivationStateResolver() {
    }

    public static boolean isActivatedForHome(
            boolean hasModernLibXposedService,
            boolean selfLoaded) {
        return hasModernLibXposedService || selfLoaded;
    }

    public static boolean hasModernLibXposedService(XposedService service) {
        if (service == null) {
            return false;
        }
        try {
            return isModernLibXposedServiceApi(service.getApiVersion());
        } catch (RuntimeException exception) {
            return false;
        }
    }

    public static boolean isModernLibXposedServiceApi(int apiVersion) {
        return apiVersion >= 101;
    }
}

File: app/src/main/java/com/dpis/module/home/HomeActivationStateResolver.kt
package com.dpis.module.home

object HomeActivationStateResolver {
    @JvmStatic
    fun isActivatedForHome(
        hasModernLibXposedService: Boolean,
        selfLoaded: Boolean,
    ): Boolean = hasModernLibXposedService || selfLoaded

    /** A disabled detector deliberately presents Home as enabled for known card false positives. */
    @JvmStatic
    fun isActivatedForHome(
        detectionEnabled: Boolean,
        hasModernLibXposedService: Boolean,
        selfLoaded: Boolean,
    ): Boolean = !detectionEnabled || isActivatedForHome(hasModernLibXposedService, selfLoaded)

    fun hasModernLibXposedService(readApiVersion: (() -> Int)?): Boolean =
        readApiVersion != null && isModernLibXposedServiceApi(readApiVersion)

    internal fun isModernLibXposedServiceApi(readApiVersion: () -> Int): Boolean = try {
        isModernLibXposedServiceApi(readApiVersion())
    } catch (_: RuntimeException) {
        false
    }

    @JvmStatic
    fun isModernLibXposedServiceApi(apiVersion: Int): Boolean = apiVersion >= 101
}
